include/pybind11/stl_bind.h

```diff
@@ -61,10 +61,21 @@ struct is_comparable<
 /* For a vector/map data structure, recursively check the value type
    (which is std::pair for maps) */
 template <typename T>
-struct is_comparable<T, enable_if_t<container_traits<T>::is_vector>> {
+struct is_comparable<T,
+                     enable_if_t<container_traits<T>::is_vector &&
+                                 // Avoid this instantiation if the type is recursive
+                                 negation<std::is_same<T, typename T::value_type>>::value>> {
```

Comment:
The comment above the specialization mentions map types, but the `enable_if` condition requires vector types. 
Hence, I only added the simple check here.